### ISSUE TITLE
termination error should possibly be thrown later than it currently is

### ISSUE THREAD

**Issue by mb706** [mlr-org/mlr3tuning#210]
```r
ps <- ParamSet$new(list(
  ParamDbl$new("cp", lower = 0, upper = 1),
  ParamInt$new("minsplit", lower = 1, upper = 20)))
ti <- TuningInstance$new(tsk("pima"),
  lrn("classif.rpart", predict_type = "prob"), rsmp("cv"),
  msrs(c("classif.auc", "classif.logloss")),
  ps, term("evals", n_evals = 1))
ti$eval_batch(data.table(cp = 0, minsplit = 1))
```
throws an error
```r
Error: TuningInstance (tsk:pima, lrn:classif.rpart, term:<TerminatorEvals>) terminated
```
we should consider throwing this error not after the last evaluation was performed, but before an evaluation is tried _after_ the process has terminated. I.e.
```r
ti <- TuningInstance$new(..., term("evals", n_evals = 1))
ti$eval_batch(data.table(cp = 0, minsplit = 1))  # should work
ti$eval_batch(data.table(cp = 1, minsplit = 2))  # should throw an error
```

**Comment by berndbischl** [comment]
in the TuningInstance$eval_batch we check before and after the evaluation for termination.
the check AFTER the eval is wrong, isnt it? is there any downside removing that?

**Comment by jakob-r** [comment]
The only case I can think of atm., is if we evaluate batches A and B in parallel.
Now A finishes earlier and already satisfies the termination criterion. What should happen with the results of batch B?
The problem might be in `self$terminator$is_terminated(self)`. Of course it should not be a problem if batch A was able to terminate itself. But it might not be feasible to add results to a terminated instance. In this case B could throw an error if `self$terminator$is_terminated(self_but_without_my_results)`. So we probably still have to check twice for termination: Once before we add results to the archive and  oncebefore we start a calculation

**Comment by berndbischl** [comment]
> The only case I can think of atm., is if we evaluate batches A and B in parallel.
> Now A finishes earlier and already satisfies the termination criterion. What should happen with the results of batch B

what you describe is currently unsupported and out of scope? batches exist so we parallelize INSIDE of the batches? and the docs clearly say that?

**Comment by berndbischl** [comment]
so all batches are spawned sequentially on the master

**Comment by berndbischl** [comment]
PR
https://github.com/mlr-org/mlr3tuning/pull/211
